File: internal/auth/service.go
package auth

import (
	"ResendIt/internal/security"
	"ResendIt/internal/user"
	"errors"
)

type Service struct {
	repo *Repository
}

func NewService(r *Repository) *Service {
	return &Service{repo: r}
}

func (s *Service) Login(username, password string) (string, error) {
	u, err := s.repo.FindByUsername(username)

	if errors.Is(err, user.ErrUserNotFound) {
		// Prevent user enumeration by returning a generic error message
		return "", ErrInvalidCredentials
	} else if err != nil {
		return "", err
	}

	if !security.CheckPassword(password, u.PasswordHash) {
		return "", ErrInvalidCredentials
	}

	return GenerateJWT(u.Username, u.Role)
}

File: src/server/auth/service.ts
import { checkPassword } from '../security/password';
import { UserNotFoundError, type UserService } from '../users/service';
import { generateJwt } from './jwt';

export class InvalidCredentialsError extends Error {
  constructor() {
    super('invalid credentials');
    this.name = 'InvalidCredentialsError';
  }
}

export class AuthService {
  constructor(private readonly users: UserService) {}

  async login(username: string, password: string): Promise<string> {
    const user = await this.users.findByUsername(username).catch((err) => {
      // A missing user and a wrong password must be indistinguishable.
      if (err instanceof UserNotFoundError) throw new InvalidCredentialsError();
      throw err;
    });

    if (!(await checkPassword(password, user.passwordHash))) {
      throw new InvalidCredentialsError();
    }

    return generateJwt(String(user.id), user.username, user.role);
  }
}
